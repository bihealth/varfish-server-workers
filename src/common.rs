//! Common functionality.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    ops::Range,
    path::Path,
};

use byte_unit::Byte;
use clap_verbosity_flag::{InfoLevel, Verbosity};

use clap::Parser;
use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};
use hgvs::static_data::Assembly;
use indexmap::IndexMap;

/// Commonly used command line arguments.
#[derive(Parser, Debug)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: Verbosity<InfoLevel>,
}

/// Helper to print the current memory resident set size via `tracing`.
pub fn trace_rss_now() {
    let me = procfs::process::Process::myself().unwrap();
    let page_size = procfs::page_size();
    tracing::debug!(
        "RSS now: {}",
        Byte::from_bytes((me.stat().unwrap().rss * page_size) as u128).get_appropriate_unit(true)
    );
}

/// Definition of canonical chromosome names.
pub const CHROMS: &[&str] = &[
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17",
    "18", "19", "20", "21", "22", "X", "Y", "M",
];

/// Build mapping of chromosome names to chromosome counts.
pub fn build_chrom_map() -> IndexMap<String, usize> {
    let mut result = IndexMap::new();
    for (i, &chrom_name) in CHROMS.iter().enumerate() {
        result.insert(chrom_name.to_owned(), i);
        result.insert(format!("chr{chrom_name}").to_owned(), i);
    }
    result.insert("x".to_owned(), 22);
    result.insert("y".to_owned(), 23);
    result.insert("chrx".to_owned(), 22);
    result.insert("chry".to_owned(), 23);
    result.insert("mt".to_owned(), 24);
    result.insert("m".to_owned(), 24);
    result.insert("chrmt".to_owned(), 24);
    result.insert("chrm".to_owned(), 24);
    result.insert("MT".to_owned(), 24);
    result.insert("chrMT".to_owned(), 24);
    result
}

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently opena  file with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// Compute reciprocal overlap between two ranges.
pub fn reciprocal_overlap(lhs: Range<i32>, rhs: Range<i32>) -> f32 {
    let lhs_b = lhs.start;
    let lhs_e = lhs.end;
    let rhs_b = rhs.start;
    let rhs_e = rhs.end;
    let ovl_b = std::cmp::max(lhs_b, rhs_b);
    let ovl_e = std::cmp::min(lhs_e, rhs_e);
    if ovl_b >= ovl_e {
        0f32
    } else {
        let ovl_len = (ovl_e - ovl_b) as f32;
        let x1 = ovl_len / (lhs_e - lhs_b) as f32;
        let x2 = ovl_len / (rhs_e - rhs_b) as f32;
        x1.min(x2)
    }
}

/// Helper to convert ENSEMBL and RefSeq gene ID to u32.
pub fn numeric_gene_id(raw_id: &str) -> Result<u32, anyhow::Error> {
    let clean_id = if raw_id.starts_with("ENSG") {
        // Strip "ENSG" prefix and as many zeroes as follow
        raw_id
            .chars()
            .skip("ENSG".len())
            .skip_while(|c| *c == '0')
            .collect()
    } else {
        raw_id.to_owned()
    };

    clean_id
        .parse::<u32>()
        .map_err(|e| anyhow::anyhow!("could not parse gene id {:?}: {}", &clean_id, &e))
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

/// Select the genome release to use.
#[derive(
    clap::ValueEnum,
    Clone,
    Copy,
    Debug,
    strum::Display,
    PartialEq,
    Eq,
    enum_map::Enum,
    PartialOrd,
    Ord,
    Hash,
)]
pub enum GenomeRelease {
    // GRCh37 / hg19
    #[strum(serialize = "GRCh37")]
    Grch37,
    /// GRCh38 / hg38
    #[strum(serialize = "GRCh38")]
    Grch38,
}

impl GenomeRelease {
    pub fn name(&self) -> String {
        match self {
            GenomeRelease::Grch37 => String::from("GRCh37"),
            GenomeRelease::Grch38 => String::from("GRCh38"),
        }
    }
}

impl From<GenomeRelease> for Assembly {
    fn from(val: GenomeRelease) -> Self {
        match val {
            GenomeRelease::Grch37 => Assembly::Grch37p10,
            GenomeRelease::Grch38 => Assembly::Grch38,
        }
    }
}

impl From<Assembly> for GenomeRelease {
    fn from(assembly: Assembly) -> Self {
        match assembly {
            Assembly::Grch37 | Assembly::Grch37p10 => GenomeRelease::Grch37,
            Assembly::Grch38 => GenomeRelease::Grch38,
        }
    }
}

impl std::str::FromStr for GenomeRelease {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        if s.starts_with("grch37") {
            Ok(GenomeRelease::Grch37)
        } else if s.starts_with("grch38") {
            Ok(GenomeRelease::Grch38)
        } else {
            Err(anyhow::anyhow!("Unknown genome release: {}", s))
        }
    }
}

/// The version of `varfish-server-worker` package.
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn numeric_gene_id_simple() -> Result<(), anyhow::Error> {
        assert_eq!(1, numeric_gene_id("ENSG0000000001")?);
        assert_eq!(1, numeric_gene_id("ENSG1")?);
        assert_eq!(1, numeric_gene_id("1")?);

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[test]
    fn trace_rss_now_smoke() {
        super::trace_rss_now();
    }

    #[test]
    fn build_chrom_map_snapshot() {
        let map = super::build_chrom_map();
        insta::assert_yaml_snapshot!(map);
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_read_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let mut f = super::open_read_maybe_gz(if is_gzip {
            "tests/common/test.txt.gz"
        } else {
            "tests/common/test.txt"
        })?;

        let mut buf = String::new();
        f.read_to_string(&mut buf)?;

        insta::assert_snapshot!(&buf);

        Ok(())
    }

    #[rstest::rstest]
    #[case(0..10, 0..10, 1.0)]
    #[case(0..10, 5..15, 0.5)]
    #[case(5..15, 0..10, 0.5)]
    #[case(0..10, 10..20, 0.0)]
    #[case(0..2, 0..10, 0.2)]
    #[case(0..10, 0..2, 0.2)]
    fn reciprocal_overlap(
        #[case] lhs: std::ops::Range<i32>,
        #[case] rhs: std::ops::Range<i32>,
        #[case] expected: f32,
    ) {
        let actual = super::reciprocal_overlap(lhs, rhs);
        assert!(float_cmp::approx_eq!(f32, expected, actual, ulps = 2))
    }

    #[rstest::rstest]
    #[case("ENSG0000000142", 142)]
    #[case("42", 42)]
    fn numeric_gene_id(#[case] raw_id: &str, #[case] expected: u32) {
        let actual = super::numeric_gene_id(raw_id).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "GRCh37")]
    #[case(crate::common::GenomeRelease::Grch38, "GRCh38")]
    fn genome_release_name(#[case] release: super::GenomeRelease, #[case] expected: &str) {
        assert_eq!(expected, release.name());
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn assembly_from_genome_release(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: hgvs::static_data::Assembly = release.into();

        assert_eq!(res, assembly);

        Ok(())
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37
    )]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn genome_release_from_assembly(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = assembly.into();

        assert_eq!(res, release);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "grch37")]
    #[case(crate::common::GenomeRelease::Grch38, "grch38")]
    fn genome_relese_from_str(
        #[case] release: super::GenomeRelease,
        #[case] s: &str,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = s.parse()?;

        assert_eq!(res, release);

        Ok(())
    }
}
