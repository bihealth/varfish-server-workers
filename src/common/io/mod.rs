//! Common, IO-related code.

pub mod std;
