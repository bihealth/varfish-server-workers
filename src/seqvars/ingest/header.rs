use std::collections::HashSet;

use noodles_vcf as vcf;

use crate::common::GenomeRelease;

/// Enumeration for the known variant callers.
#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
pub enum VariantCaller {
    GatkHaplotypeCaller {
        version: String,
    },
    GatkUnifiedGenotyper {
        version: String,
    },
    Glnexus {
        version: String,
        config_name: Option<String>,
    },
    Dragen {
        version: String,
    },
    Other,
}

impl VariantCaller {
    /// Return a string with the name of the variant caller for the VCF header string.
    fn name(&self) -> &'static str {
        match self {
            VariantCaller::GatkHaplotypeCaller { .. } => "GatkHaplotypeCaller",
            VariantCaller::GatkUnifiedGenotyper { .. } => "GatkUnifiedGenotyper",
            VariantCaller::Dragen { .. } => "Dragen",
            VariantCaller::Glnexus { .. } => "Glnexus",
            VariantCaller::Other => "Other",
        }
    }
}

impl VariantCaller {
    pub fn guess(header: &vcf::Header) -> Option<Self> {
        use vcf::header::record::value::collection::Collection;

        let mut glnexus_version: Option<String> = None;
        let mut glnexus_config_name: Option<String> = None;

        for (other, collection) in header.other_records() {
            if ["GATKCommandLine", "DRAGENCommandLine"]
                .iter()
                .any(|k| other.as_ref().starts_with(k))
            {
                if let Collection::Structured(map) = collection {
                    for (key, values) in map.iter() {
                        match (key.as_str(), values.other_fields().get("Version").cloned()) {
                            ("HaplotypeCaller", Some(version)) => {
                                return Some(VariantCaller::GatkHaplotypeCaller { version })
                            }
                            ("UnifiedGenotyper", Some(version)) => {
                                return Some(VariantCaller::GatkUnifiedGenotyper { version })
                            }
                            ("dragen", Some(version)) => {
                                return Some(VariantCaller::Dragen { version })
                            }
                            _ => (),
                        }
                    }
                }
            } else if other.as_ref().starts_with("GLnexusVersion") {
                if let Collection::Unstructured(values) = collection {
                    glnexus_version = Some(values[0].clone());
                }
            } else if other.as_ref().starts_with("GLnexusConfigName") {
                if let Collection::Unstructured(values) = collection {
                    glnexus_config_name = Some(values[0].clone());
                }
            }
        }

        if let Some(version) = glnexus_version {
            return Some(VariantCaller::Glnexus {
                version,
                config_name: glnexus_config_name,
            });
        }

        None
    }
}

/// Add contigs for GRCh37.
fn add_contigs_37(builder: vcf::header::Builder) -> Result<vcf::header::Builder, anyhow::Error> {
    use vcf::header::record::value::map::Contig;
    use vcf::header::record::value::Map;

    let mut builder = builder;

    let specs: &[(&str, usize); 25] = &[
        ("1", 249250621),
        ("2", 243199373),
        ("3", 198022430),
        ("4", 191154276),
        ("5", 180915260),
        ("6", 171115067),
        ("7", 159138663),
        ("8", 146364022),
        ("9", 141213431),
        ("10", 135534747),
        ("11", 135006516),
        ("12", 133851895),
        ("13", 115169878),
        ("14", 107349540),
        ("15", 102531392),
        ("16", 90354753),
        ("17", 81195210),
        ("18", 78077248),
        ("19", 59128983),
        ("20", 63025520),
        ("21", 48129895),
        ("22", 51304566),
        ("X", 155270560),
        ("Y", 59373566),
        ("MT", 16569),
    ];

    for (contig, length) in specs {
        builder = builder.add_contig(
            contig
                .parse()
                .map_err(|_| anyhow::anyhow!("invalid contig: {}", contig))?,
            Map::<Contig>::builder()
                .set_length(*length)
                .insert(
                    "assembly"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: assembly"))?,
                    "GRCh37",
                )
                .insert(
                    "species"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: species"))?,
                    "Homo sapiens",
                )
                .build()?,
        );
    }

    Ok(builder)
}

/// Add contigs for GRCh38.
fn add_contigs_38(builder: vcf::header::Builder) -> Result<vcf::header::Builder, anyhow::Error> {
    use vcf::header::record::value::map::Contig;
    use vcf::header::record::value::Map;

    let mut builder = builder;

    let specs: &[(&str, usize); 25] = &[
        ("chr1", 248956422),
        ("chr2", 242193529),
        ("chr3", 198295559),
        ("chr4", 190214555),
        ("chr5", 181538259),
        ("chr6", 170805979),
        ("chr7", 159345973),
        ("chr8", 145138636),
        ("chr9", 138394717),
        ("chr10", 133797422),
        ("chr11", 135086622),
        ("chr12", 133275309),
        ("chr13", 114364328),
        ("chr14", 107043718),
        ("chr15", 101991189),
        ("chr16", 90338345),
        ("chr17", 83257441),
        ("chr18", 80373285),
        ("chr19", 58617616),
        ("chr20", 64444167),
        ("chr21", 46709983),
        ("chr22", 50818468),
        ("chrX", 156040895),
        ("chrY", 57227415),
        ("chrM", 16569),
    ];

    for (contig, length) in specs {
        builder = builder.add_contig(
            contig
                .parse()
                .map_err(|_| anyhow::anyhow!("invalid contig: {}", contig))?,
            Map::<Contig>::builder()
                .set_length(*length)
                .insert(
                    "assembly"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: assembly"))?,
                    "GRCh38",
                )
                .insert(
                    "species"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: species"))?,
                    "Homo sapiens",
                )
                .build()?,
        );
    }

    Ok(builder)
}

/// Generate the output header from the input header.
pub fn build_output_header(
    input_header: &vcf::Header,
    pedigree: &Option<mehari::ped::PedigreeByName>,
    id_mapping: &Option<indexmap::IndexMap<String, String>>,
    genomebuild: GenomeRelease,
    file_date: &str,
    case_uuid: &uuid::Uuid,
    worker_version: &str,
) -> Result<vcf::Header, anyhow::Error> {
    use vcf::header::record::value::{
        map::{info::Type, Filter, Format, Info},
        Map,
    };
    use vcf::header::Number;
    use vcf::record::genotypes::keys::key;

    let builder = vcf::Header::builder()
        .insert(
            "fileDate".parse()?,
            vcf::header::record::Value::from(file_date),
        )?
        .add_filter("PASS", Map::<Filter>::new("All filters passed"))
        .add_info(
            "gnomad_exomes_an".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of alleles in gnomAD exomes",
            ),
        )
        .add_info(
            "gnomad_exomes_hom".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of hom. alt. carriers in gnomAD exomes",
            ),
        )
        .add_info(
            "gnomad_exomes_het".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of het. alt. carriers in gnomAD exomes",
            ),
        )
        .add_info(
            "gnomad_exomes_hemi".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of hemi. alt. carriers in gnomAD exomes",
            ),
        )
        .add_info(
            "gnomad_genomes_an".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of alleles in gnomAD genomes",
            ),
        )
        .add_info(
            "gnomad_genomes_hom".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of hom. alt. carriers in gnomAD genomes",
            ),
        )
        .add_info(
            "gnomad_genomes_het".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of het. alt. carriers in gnomAD genomes",
            ),
        )
        .add_info(
            "gnomad_genomes_hemi".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of hemi. alt. carriers in gnomAD genomes",
            ),
        )
        .add_info(
            "helix_an".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of alleles in HelixMtDb",
            ),
        )
        .add_info(
            "helix_hom".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of hom. alt. carriers in HelixMtDb",
            ),
        )
        .add_info(
            "helix_het".parse()?,
            Map::<Info>::new(
                Number::Count(1),
                Type::Integer,
                "Number of het. alt. carriers in HelixMtDb",
            ),
        )
        .add_info(
            "ANN".parse()?,
            Map::<Info>::new(
                Number::Unknown,
                Type::String,
                "Functional annotations: 'Allele | Annotation | Annotation_Impact | \
                Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | \
                HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | \
                AA.pos / AA.length | Distance | ERRORS / WARNINGS / INFO'",
            ),
        )
        .add_format(key::READ_DEPTHS, Map::<Format>::from(&key::READ_DEPTHS))
        .add_format(key::READ_DEPTH, Map::<Format>::from(&key::READ_DEPTH))
        .add_format(
            key::CONDITIONAL_GENOTYPE_QUALITY,
            Map::<Format>::from(&key::CONDITIONAL_GENOTYPE_QUALITY),
        )
        .add_format(key::GENOTYPE, Map::<Format>::from(&key::GENOTYPE))
        .add_format(key::PHASE_SET, Map::<Format>::from(&key::PHASE_SET));

    let mut builder = match genomebuild {
        GenomeRelease::Grch37 => add_contigs_37(builder),
        GenomeRelease::Grch38 => add_contigs_38(builder),
    }
    .map_err(|e| anyhow::anyhow!("problem adding contigs: {}", e))?;

    if let Some(pedigree) = pedigree {
        let ped_idv = pedigree
            .individuals
            .iter()
            .map(|(name, _)| name.clone())
            .collect::<HashSet<_>>();
        let input_idv = input_header
            .sample_names()
            .iter()
            .cloned()
            .collect::<HashSet<_>>();
        if let Some(id_mapping) = id_mapping {
            let src_idv = id_mapping.keys().cloned().collect::<HashSet<_>>();
            if !src_idv.eq(&input_idv) {
                anyhow::bail!(
                    "mapping source individuals = {:?} != input individuals: {:?}",
                    &src_idv,
                    &input_idv
                )
            }
            let dst_idv = id_mapping.values().cloned().collect::<HashSet<_>>();
            if !dst_idv.eq(&ped_idv) {
                anyhow::bail!(
                    "mapping destination individuals = {:?} != pedigree individuals: {:?}",
                    &dst_idv,
                    &ped_idv
                )
            }
        } else if !ped_idv.eq(&input_idv) {
            anyhow::bail!(
                "pedigree individuals = {:?} != input individuals: {:?}",
                &ped_idv,
                &input_idv
            )
        }

        let mut sample_names = Vec::new();
        for src_name in input_header.sample_names() {
            let dst_name = if let Some(id_mapping) = id_mapping {
                id_mapping.get(src_name).expect("checked above")
            } else {
                src_name
            };

            let i = pedigree.individuals.get(dst_name).expect("checked above");
            if input_header.sample_names().contains(src_name) {
                sample_names.push(dst_name.clone());
            }

            // Add SAMPLE entry.
            builder = builder.insert(
                "SAMPLE".parse()?,
                noodles_vcf::header::record::Value::Map(
                    i.name.clone(),
                    Map::<Other>::builder()
                        .insert(
                            "Sex".parse()?,
                            mehari::annotate::strucvars::vcf_header::sex_str(i.sex),
                        )
                        .insert(
                            "Disease".parse()?,
                            mehari::annotate::strucvars::vcf_header::disease_str(i.disease),
                        )
                        .build()?,
                ),
            )?;

            // Add PEDIGREE entry.
            let mut map_builder = Map::<Other>::builder();
            if let Some(father) = i.father.as_ref() {
                map_builder = map_builder.insert("Father".parse()?, father.clone());
            }
            if let Some(mother) = i.mother.as_ref() {
                map_builder = map_builder.insert("Mother".parse()?, mother.clone());
            }
            builder = builder.insert(
                "PEDIGREE".parse()?,
                noodles_vcf::header::record::Value::Map(i.name.clone(), map_builder.build()?),
            )?;
        }

        sample_names.sort();
        builder = builder.set_sample_names(sample_names.into_iter().collect());
    } else {
        for name in input_header.sample_names() {
            let name = if let Some(id_mapping) = id_mapping {
                id_mapping
                    .get(name)
                    .ok_or_else(|| anyhow::anyhow!("mapping given but sample {} not found", name))?
                    .clone()
            } else {
                name.clone()
            };
            builder = builder.add_sample_name(name);
        }
    }

    use vcf::header::record::value::map::Other;

    let orig_caller = VariantCaller::guess(input_header)
        .ok_or_else(|| anyhow::anyhow!("unable to guess original variant caller"))?;

    let builder = builder
        .insert(
            "x-varfish-case-uuid".parse()?,
            vcf::header::record::Value::String(case_uuid.to_string()),
        )?
        .insert(
            "x-varfish-version".parse()?,
            vcf::header::record::Value::Map(
                String::from("varfish-server-worker"),
                Map::<Other>::builder()
                    .insert("Version".parse()?, worker_version)
                    .build()?,
            ),
        )?;

    let builder = match &orig_caller {
        VariantCaller::GatkHaplotypeCaller { version }
        | VariantCaller::GatkUnifiedGenotyper { version }
        | VariantCaller::Dragen { version } => builder.insert(
            "x-varfish-version".parse()?,
            vcf::header::record::Value::Map(
                String::from("orig-caller"),
                Map::<Other>::builder()
                    .insert("Name".parse()?, orig_caller.name())
                    .insert("Version".parse()?, version)
                    .build()?,
            ),
        )?,
        VariantCaller::Glnexus {
            version,
            config_name,
        } => builder.insert(
            "x-varfish-version".parse()?,
            vcf::header::record::Value::Map(
                String::from("orig-caller"),
                Map::<Other>::builder()
                    .insert("Name".parse()?, orig_caller.name())
                    .insert("Version".parse()?, version)
                    .insert(
                        "ConfigName".parse()?,
                        config_name.clone().unwrap_or_default(),
                    )
                    .build()?,
            ),
        )?,
        VariantCaller::Other => builder.insert(
            "x-varfish-version".parse()?,
            vcf::header::record::Value::Map(
                String::from("orig-caller"),
                Map::<Other>::builder()
                    .insert("Name".parse()?, "Other")
                    .build()?,
            ),
        )?,
    };

    Ok(builder.build())
}

#[cfg(test)]
mod test {
    use mehari::ped::PedigreeByName;
    use noodles_vcf as vcf;
    use rstest::rstest;

    use super::VariantCaller;

    #[rstest]
    #[case("tests/seqvars/ingest/clair3_glnexus.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.4.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.9.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.3.7-0.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.4.4.0.0.vcf")]
    fn variant_caller_guess(#[case] path: &str) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", path.split('/').last().unwrap());

        let vcf_header = noodles_vcf::reader::Builder::default()
            .build_from_path(path)?
            .read_header()?;

        insta::assert_yaml_snapshot!(VariantCaller::guess(&vcf_header));

        Ok(())
    }

    #[rstest]
    #[case("tests/seqvars/ingest/clair3_glnexus.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.4.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.9.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.3.7-0.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.4.4.0.0.vcf")]
    fn build_output_header_37(#[case] path: &str) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", path.split('/').last().unwrap());
        let tmpdir = temp_testdir::TempDir::default();

        let pedigree = PedigreeByName::from_path(path.replace(".vcf", ".ped")).unwrap();

        let mut input_vcf_header = noodles_vcf::reader::Builder::default()
            .build_from_path(path)?
            .read_header()?;
        let output_vcf_header = super::build_output_header(
            &input_vcf_header,
            &Some(pedigree),
            &None,
            crate::common::GenomeRelease::Grch37,
            "20230421",
            &uuid::Uuid::parse_str("00000000-0000-0000-0000-000000000000").unwrap(),
            "x.y.z",
        )?;

        // Work around glnexus issue with RNC.
        if let Some(format) = input_vcf_header.formats_mut().get_mut("RNC") {
            *format.number_mut() = vcf::header::Number::Count(1);
            *format.type_mut() = vcf::header::record::value::map::format::Type::String;
        }

        let out_path = tmpdir.join("out.vcf");
        let out_path_str = out_path.to_str().expect("invalid path");
        {
            noodles_vcf::writer::Writer::new(std::fs::File::create(out_path_str)?)
                .write_header(&output_vcf_header)?;
        }

        insta::assert_snapshot!(std::fs::read_to_string(out_path_str)?);

        Ok(())
    }

    #[rstest]
    #[case("tests/seqvars/ingest/clair3_glnexus.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.4.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.9.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.3.7-0.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.4.4.0.0.vcf")]
    fn build_output_header_38(#[case] path: &str) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", path.split('/').last().unwrap());
        let tmpdir = temp_testdir::TempDir::default();

        let pedigree = PedigreeByName::from_path(path.replace(".vcf", ".ped")).unwrap();

        let mut input_vcf_header = noodles_vcf::reader::Builder::default()
            .build_from_path(path)?
            .read_header()?;
        let output_vcf_header = super::build_output_header(
            &input_vcf_header,
            &Some(pedigree),
            &None,
            crate::common::GenomeRelease::Grch38,
            "20230421",
            &uuid::Uuid::parse_str("00000000-0000-0000-0000-000000000000").unwrap(),
            "x.y.z",
        )?;

        // Work around glnexus issue with RNC.
        if let Some(format) = input_vcf_header.formats_mut().get_mut("RNC") {
            *format.number_mut() = vcf::header::Number::Count(1);
            *format.type_mut() = vcf::header::record::value::map::format::Type::String;
        }

        let out_path = tmpdir.join("out.vcf");
        let out_path_str = out_path.to_str().expect("invalid path");
        {
            noodles_vcf::writer::Writer::new(std::fs::File::create(out_path_str)?)
                .write_header(&output_vcf_header)?;
        }

        insta::assert_snapshot!(std::fs::read_to_string(out_path_str)?);

        Ok(())
    }
}
