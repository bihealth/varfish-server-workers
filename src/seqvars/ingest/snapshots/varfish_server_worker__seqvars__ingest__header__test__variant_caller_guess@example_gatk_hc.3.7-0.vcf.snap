---
source: src/seqvars/ingest/header.rs
expression: "VariantCaller::guess(&vcf_header)"
---
GatkHaplotypeCaller:
  version: 3.7-0-gcfedb67
